File: tutorial/java/Client/src/main/java/systems/zlink/tutorial/client/LegacyPlayerController.java
package systems.zlink.tutorial.client;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RestController;

import java.net.URI;

@RestController
final class LegacyPlayerController {

    @GetMapping("/player/{playerId}")
    ResponseEntity<Void> redirect(@PathVariable String playerId) {
        return ResponseEntity.status(301).location(URI.create("/players/" + playerId)).build();
    }
}

File: tutorial/kotlin/Client/src/main/kotlin/systems/zlink/tutorial/client/LegacyPlayerController.kt
package systems.zlink.tutorial.client

import java.net.URI
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RestController

@RestController
class LegacyPlayerController {

    @GetMapping("/player/{playerId}")
    fun redirect(@PathVariable playerId: String): ResponseEntity<Void> =
        ResponseEntity.status(301).location(URI.create("/players/$playerId")).build()
}
